### DISCUSSION TITLE
_source alternatives?

### DISCUSSION THREAD

**Discussion by lymanaukuleles** [jertel/elastalert2#811] [Q&A] [answered]
This is what it says in the ElastAlert guide (https://elastalert.readthedocs.io/en/latest/ruletypes.html). It’s referring to the `_source_enabled` field that you can put in your config:

`_source_enabled`: If true, ElastAlert will use `_source` to retrieve fields from documents in Elasticsearch. If false, ElastAlert will use `fields` to retrieve stored fields. Both of these are represented internally as if they came from `_source`. See https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-fields.html for more details. The fields used come from `include`, see above for more details. (Optional, boolean, default True)

So I am most curious about this line: “If false, ElastAlert will use `fields` to retrieve stored fields.” Any idea what they mean?

Actually I believe they are referring to this in the same link: 
`fields`: A list of fields to monitor for new terms. `query_key` will be used if `fields` is not set. Each entry in the list of fields can itself be a list. If a field entry is provided as a list, it will be interpreted as a set of fields that compose a composite key used for the ElasticSearch query.

So my new question is what do I add to this `fields` list? Would any field from my logs work?

**Comment by nsano-rururu** [comment, marked as answer]
As far as I can see from the source code, I think the settings are as follows, but when I try it, the error message `Error: No timestamp was found for hit. '_source_enabled' is set to false, check your mappings for stored fields` is output.
If `_source_enabled` is set to false, it doesn't seem to work properly.

```yaml
_source_enabled: false
include: ["@timestamp", "message"]
```

https://github.com/jertel/elastalert2/blob/master/elastalert/elastalert.py#L320
```python
            if not ts and not rule["_source_enabled"]:
                raise EAException(
                    "Error: No timestamp was found for hit. '_source_enabled' is set to false, check your mappings for stored fields"
                )
```

https://github.com/jertel/elastalert2/blob/master/elastalert/elastalert.py#L362
```python
        if not rule.get('_source_enabled'):
            query['stored_fields'] = rule['include']
            extra_args = {}
```

**Reply by nsano-rururu** [reply]
There may be some hints in the following pull request exchanges when added in 2015 with the original yelp / elastalert
https://github.com/Yelp/elastalert/pull/160
